Container module

zaibon opened the pull request.

zaibon committed:
add lifecyle schema + links to container runtime documentations

zaibon committed:
add lifecyle schema + links to container runtime documentations

zaibon commented:
I added some link worth reading.
I have been reading quite a bit the code of containerd see what we can leverage from them.
My conclusion on this is we have 2 way of doing it:
- We try to extract the runtime package from containerd and just integrate this one. This package is everything regarding the management of container-shim and runc. So it is the lowest level I think
- We integrate containerd itself as the container runtime. They provide a really nice client to it that make container management super easy. I believe we can plug our 0-fs in replacement of the normal image management so this is probably the easiest way.

So the decision really comes down to how much code we want to bring in. Since we want to try to make 0-OS as low level as possible I would prefer the first solution. But time is also running against us, so the second option is maybe the fastest way to victory. And over time we could strip down stuff we don't need.

muhamadazmy commented:
@zaibon 

I actually think that leverage on the full containerd is the easiest and fastest approach. And it's probably much easier to just use the client from the container module to interface with the system.

zaibon committed:
organize flist module

- split flist package into lib and daemon
- move flist daemon code to flistd directory
- add client.go that allow to get a Flister easily by other libs

zaibon committed:
add containerd client example

- the example starts a caddy container using an flist and containerd

Commit:
WIP: building the container module

muhamadazmy committed:
Merge branch 'container_module' of github.com:threefoldtech/zosv2 into container_module

muhamadazmy committed:
Update readme

muhamadazmy committed:
WIP: container modules specs

muhamadazmy committed:
WIP: container module

muhamadazmy committed:
remove binary file !

muhamadazmy committed:
WIP: add namespaces

muhamadazmy committed:
Regenrate stubs

muhamadazmy committed:
Implement delete container

muhamadazmy committed:
Handle mounts

muhamadazmy committed:
add readme

zaibon commented on modules/container.go:
```diff
@@ -0,0 +1,53 @@
+	// ns: tenant namespace
+	// name: name of container
+	// flist: flist of container
+	Run(ns, name, flist string, tags, env []string, network NetworkInfo,
```
Seems like we could wraps all these argument into a struct instead. The number of argument starts to be a bit much I think

muhamadazmy replied:
Yup, was totally thinking the same. will do it

zaibon commented on modules/container/container.go:
```diff
@@ -0,0 +1,219 @@
+	}
+	defer client.Close()
+
+	args, _ := shlex.Split(entrypoint)
```
What does this return if `entrypoint` is empty ?

muhamadazmy replied:
it will return an empty []string list

zaibon commented on modules/container/container.go:
```diff
@@ -0,0 +1,219 @@
+		return id, err
+	}
+
+	defer func() {
```
You should not defer Umount here.
You only unmount when you delete the container.

muhamadazmy replied:
we defer here only if an error occurs (container failed to start) so there will not be a delete call afterwards

zaibon replied:
I'm really not sure about this pattern.
I think defer does evaluate the the value of err when you call defer, not when the function deferred is called.
https://golang.org/doc/effective_go.html#defer
So this means here err will always be false at this point, thus the Umount will never bee called even in the case there is an error when you create the container.

muhamadazmy replied:
I don't believe so, i used this pattern before many times and it works as expected. but let me test again to be sure.

muhamadazmy replied:
Notice also i use a named return. so (err) in all function reference the same err object, hence once the err is actually populated to value it will be detected by error.

muhamadazmy replied:
https://play.golang.org/p/_sKT9UBDwdV

zaibon commented on modules/container/container.go:
```diff
@@ -0,0 +1,219 @@
+		return id, err
+	}
+
+	defer func() {
```
Should not defer Delete here.

muhamadazmy replied:
this only happens if an error occurs, we make sure we delete the container if the task start fails (then followed by unmount)

zaibon replied:
Same remark here

zaibon commented on modules/container/container.go:
```diff
@@ -0,0 +1,219 @@
+	}()
+
+	logs := path.Join("/var/log", ns)
+	os.MkdirAll(logs, 0755)
```
need to check the error return by `MkdirAll`

muhamadazmy replied:
Yup, will also move it to the module start up code.

zaibon commented on modules/container/container.go:
```diff
@@ -0,0 +1,219 @@
+		}
+	}()
+
+	logs := path.Join("/var/log", ns)
```
Not sure we want to writes logs into `/var/log`
My idea is that a module will have a dedicated root directory under which it can do whatever.
This makes it easy to managed the storage used by each module and avoid spreading files all over the tree

So I would create a log directory under the root directory of the container module and use this instead

muhamadazmy replied:
Yup, makes sense. will change that.

zaibon commented on modules/container/container.go:
```diff
@@ -0,0 +1,219 @@
+			trials--
+			select {
+			case <-exitC:
+				break loop
```
Doesn't this make the flow go to the loop label and thus restart the loop?

muhamadazmy replied:
No, it's not a `goto` statement. `break <label>` will break out of the loop labeled `loop` it's used to break nested loops or in this case a select block.

muhamadazmy replied:
https://stackoverflow.com/questions/11104085/in-go-does-a-break-statement-break-from-a-switch-select

zaibon commented on modules/container/container.go:
```diff
@@ -0,0 +1,219 @@
+		return err
+	}
+
+	return syscall.Unmount(spec.Root.Path, syscall.MNT_DETACH)
```
```suggestion
	return c.flister.Umount(spec.Root.Path)
```

muhamadazmy replied:
why u wanna drop the `DETACH` it will make the unmount. this will allow the system to actually perform the unmount once the files are not used anymore. instead of getting an  (filesystem busy) error

zaibon replied:
Look again, I'm not doing a `syscall.Unmount`, but call the `Umount` method of the flister module.

zaibon commented on modules/flist/flist.go:
```diff
@@ -45,6 +45,10 @@ func New(root string) modules.Flister {
 		root = defaultRoot
 	}
 
+	if err := os.MkdirAll(root, 0770); err != nil {
```
nice catch :+1:

muhamadazmy committed:
Apply review comments

Commit:
Update modules/container/container.go

Co-Authored-By: Christophe de Carvalho <zaibon@users.noreply.github.com>

zaibon commented:
Reading the code a second time, I wonder if we should include the flister module in this module at all.
Instead of passing the URL fo the flist to this module, shoulnd't we instead just pass the path root the rootfs ?
So we keep the separation of concern between module and doesn't introduce dependencies between low level module.

The component responsible to watch the reservation can then use both container and flist module to provision the container.
This also make it easier to handle error, cause now if the flist fail to unmount for some reason at the end of the Delete method. The method will return an error. And we'll endup in a half deleted state hard to fix.

While if the reservation watcher manage both flist and container it can have a better logic regarding failure at different level.

muhamadazmy committed:
Implement inspect

Apply some review comments

maxux committed:
Update container README with new arguments

maxux committed:
Update examples

maxux committed:
Update example files name

zaibon committed:
container module: decouple flist and container modules

container module has no notion of flist, only paths
this allow to remove the dependency of flist module on the container
module

zaibon committed:
Merge branch 'master' into container_module

zaibon committed:
container module: add support for coreX in the container

- change entrypoint to coreX binary
- mount flist as /sandbox and make coreX chroot into /sandbox at start

zaibon committed:
add contctl debugging tool
